tests/scenarios.go

```diff
@@ -104,13 +125,106 @@ func LoadRequestsUndo(file string) map[string]UndoAction {
 	return value
 }
 
+// LoadGivenSteps load undo configuration.
+func LoadGivenSteps(file string) []GivenStep {
+	f, err := os.Open(file)
+	if err != nil {
+		panic(err)
+	}
+	defer f.Close()
+
+	byteValue, _ := ioutil.ReadAll(f)
+
+	var value []GivenStep
+	json.Unmarshal(byteValue, &value)
+	return value
+}
+
 // SetRequestsUndo sets map with undo function for each request.
 func SetRequestsUndo(ctx gobdd.Context, value map[string]UndoAction) {
 	ctx.Set(ctxRequestsUndoKey{}, value)
 }
 
+// RegisterSuite adds step implementation.
+func (s GivenStep) RegisterSuite(suite *gobdd.Suite) {
+	given := func(t gobdd.StepTest, ctx gobdd.Context) {
+		// get an instance of Client
+		clientInterface, err := ctx.Get(clientKey{})
+		if err != nil {
+			t.Fatalf("could not find client: %v", err)
+		}
+
+		// Enable unstable operation
+		configInterface := reflect.Indirect(clientInterface.(reflect.Value)).Addr().MethodByName("GetConfig").Call(nil)[0]
+		configInterface.MethodByName("SetUnstableOperationEnabled").Call([]reflect.Value{
+			reflect.ValueOf(s.OperationID), reflect.ValueOf(true),
+		})
+
+		// find API service based on undo tag value: `client.{{ undo.Tag }}Api`
+		tag := strings.Replace(s.Tag, " ", "", -1) + "Api"
+		api := reflect.Indirect(clientInterface.(reflect.Value)).FieldByName(tag)
+		if !api.IsValid() {
+			t.Fatalf("invalid API name %sApi", tag)
+		}
+
+		// get method from the service API: `client.{{ undo.Tag }}Api.{{ undo.Undo.OperationID }}`
+		operation := api.MethodByName(s.OperationID)
+		if !operation.IsValid() {
+			t.Fatalf("invalid method name %s", s.OperationID)
+		}
+
+		// Assemble method arguments
+		in := make([]reflect.Value, operation.Type().NumIn())
+		// first argument is always context.Context
+		in[0] = reflect.ValueOf(GetCtx(ctx))
+		for i := 1; i < operation.Type().NumIn(); i++ {
+			in[i] = s.Parameters[i-1].Resolve(ctx, operation.Type().In(i))
+		}
+		request := operation.Call(in)[0]
+
+		// Call builder pattern
+		for _, p := range s.Parameters {
+			name := ToVarName(p.Name)
+			method := request.MethodByName(name)
+			if method.IsValid() {
+				v := p.Resolve(ctx, method.Type().In(0))
+				request = method.Call([]reflect.Value{v})[0]
+			}
+		}
+
+		undo, err := GetRequestsUndo(ctx, s.OperationID)
```

Comment:
Can you add some more comments to this section? I'm not fully clear on whats happening here, it looks like there's some assumptions about accessing -1 index + the usage of "00-given" keys?